### DISCUSSION TITLE
Power  configuration for burning 27C256/512 on TommyProm32 ?

### DISCUSSION THREAD

**Discussion by ABelliqueux** [TomNisbet/TommyPROM#75] [Q&A]
I saw that the 27C256 is supported via `PromDevice27` and needs a 6.25VCC/12.75VPP

Has anyone a picture of the physical setup with jumpers and shunts in place on a Tommyprom32 so that I don't make a mistake when wiring them ?

**Comment by TomNisbet** [comment]
You can find a picture of an M27C256 here: https://tomnisbet.github.io/TommyPROM/docs/tommyprom32-pcb#m27c256
It has Vpp on pin 1 and Vcc on pin 28.

What is the complete part number for your chip?

**Reply by ABelliqueux** [reply]
Sorry I didn't see your answer yesterday. Thanks for the link, that's exactly what I was looking for. The part number is ST M27C512-45XF1.
These are  CDIP28 512 Kbit 64K x8 UV EPROM, 

Here is the datasheet : [m27c512.pdf](https://github.com/user-attachments/files/17025158/m27c512.pdf)

Vcc/Vpp seem to be the same as M27C256 (i.e ; 6,25/12,75), but Vpp is on **pin 22** (Also G̅) on the **512** and **pin 1** on the **256**. 

Does G̅ being on the same pin as Vpp have any impact ?

**Comment by ABelliqueux** [comment]
So I got a chance to try flashing a 27C512 and I'm getting a write failed error.

I suspect this cannot work as is with `PromDevice27` for a M27C512, as the pinout is a bit different : 

# Pinouts

On the **M27C256**,  Vpp and G are on two distinct pins ( 1 and 22) :  

![2024-10-06-125013_342x450_scrot](https://github.com/user-attachments/assets/4e9bbf51-2e30-4a5e-9dd6-96cd704a8706)

Whereas on the **M27C512**, Vpp and G are on the same pin (22)

![2024-10-06-125025_707x467_scrot](https://github.com/user-attachments/assets/410737c3-fc0b-451a-9b20-7383768cc239)

# Operating modes

![c256](https://github.com/user-attachments/assets/aab31f21-7b4d-425a-b90f-f3d54854b4e7)

![2024-10-06-125004_690x355_scrot](https://github.com/user-attachments/assets/157c7892-e01e-4023-92f6-63e444794016)

Specifically, the 'Programming' section of each chip differs on this point :  

> The M27C256B is in the programming mode when VPP input is at 12.75V, G is at VIH and E is pulsed to VIL.

versus :  

> The M27C512 is in the programming mode when VPP input is at 12.75V and E is pulsed to VIL.

The 'Program Verify' also differs :  

For the 27c256 : 

> The verify is accomplished with G at VIL, E at VIH, VPP at 12.75V and VCC at 6.25V.

For the 27c512: 

> The verify is accomplished with G at VIL. Data should be verified with tELQV after the falling edge of E.

# Wiring 

Vpp 1 is wired to OE and Vpp 2 is wired to A17 on the PCB.

![27c512](https://github.com/user-attachments/assets/9fdde69c-7741-41de-a1ea-cb867a9dc491)

## Read mode 

Jumpers are off (VPP Select ==  low) on Vpp select, getting 4.7V on Vpp out 1 and 2 : 

![po5mOe0rgG4bTve9](https://github.com/user-attachments/assets/698cb6d4-7def-4ec3-9bcf-22ea570f536e)

## Write mode

Jumpers are on (VPP Select == high) on Vpp select, getting 12.6V on Vpp 1 and 6.4V on Vpp 2 : 

![GRfTkmm1kpWNs4qM](https://github.com/user-attachments/assets/2f7d20cf-d8ab-4f4b-8f0b-355c9e2775dc)



@TomNisbet Do you have an idea as to what could be wrong ?

Should I open  an issue ?

**Comment by ABelliqueux** [comment]
I just [noticed this](https://tomnisbet.github.io/TommyPROM/docs/prom-device-27) in the doc : 

> Unlike the 257 chips, the W27C512 does not have a dedicated pin for the programming voltage and instead uses 12V on the OE pin to place the chip in programming mode. The verify operation requires that the OE pin be switched to LOW and there is no hardware support for this, so the current code supports the 512 chip by doing a single write cycle with no verify.

This looks similar to what the M27C512 does looking at the datasheets. 

Does this mean using a command like `F0 F 42` will actually work, although returning `ERROR: Write Failed` ?

![UkW3DUARxGFkyacZ](https://github.com/user-attachments/assets/0b5ba4e0-9312-4941-baf0-86e5e04cea38)
